filewatch/filewatcher.go

```diff
@@ -0,0 +1,152 @@
+	ConfigMap map[string]string
+
+	// TriggerChannel is a channel
+	TriggerChannel chan<- struct{}
+
+	// ErrChan is a channel that is used to receive errors.
+	ErrChan <-chan error
+
+	callbackFunc func() error
+
+	closeFunc func() error
+)
+
+type watcherType string
+
+const (
+	watcherTypeJSON watcherType = "json"
+	watcherTypeYAML watcherType = "yaml"
+)
+
+// A FileWatcher can be used to observe a file on the filesystem.
+// The Watcher will detect changes in the file, update the global ConfigMap in the application
+// and send a struct{} to the given TriggerChannels which can be used for example to implement a
+// connection-renewal for the desired technology instance.
+type FileWatcher struct {
+	Loader
+	closeFunc
+	watcherType
+	errChan         <-chan error
+	prefix          string
+	path            string
+	triggerChannels []TriggerChannel
+}
+
+// Option is a function that can be used to configure the FileWatcher.
+type Option func(*FileWatcher)
+
+// WithPrefix is a function that can be used to set the prefix for the file paths.
+func WithPrefix(prefix string) Option {
+	return func(f *FileWatcher) {
+		f.prefix = prefix
+	}
+}
+
+// WithTriggerChannels is a function that can be used to set the TriggerChannels for the FileWatcher.
+func WithTriggerChannels(triggerChannels ...TriggerChannel) Option {
+	return func(f *FileWatcher) {
+		f.triggerChannels = triggerChannels
+	}
+}
+
+// NewJSONFileWatcher creates a new File-Watcher that observes json files.
+// The prefix is optional and can be left empty.
+// Setting the Prefix is useful in case you have multiple Watchers observing multiple files,
+// which contain the same keys. The prefix will be added to the key in the global ConfigMap.
+func NewJSONFileWatcher(path string, loader Loader, options ...Option) *FileWatcher {
+	fw := &FileWatcher{
+		watcherType: watcherTypeJSON,
+		Loader:      loader,
+		path:        path,
+	}
+
+	for i := range options {
+		options[i](fw)
+	}
+
+	return fw
+}
+
+// NewYAMLFileWatcher creates a new File-Watcher that observes yaml files.
+// Setting the Prefix is useful in case you have multiple Watchers observing multiple files,
+// which contain the same keys. The prefix will be added to the key in the global ConfigMap.
+func NewYAMLFileWatcher(path string, loader Loader, options ...Option) *FileWatcher {
+	fw := &FileWatcher{
+		watcherType: watcherTypeYAML,
+		Loader:      loader,
+		path:        path,
+	}
+
+	for i := range options {
+		options[i](fw)
+	}
+
+	return fw
+}
+
+// ErrLoaderNotSet is returned when no loader is specified.
+var ErrLoaderNotSet = errors.New("no loader is specified")
+
+// Start starts the file-watcher. The config parameter is the ConfigMap
+// that will be updated when the file-watcher detects changes.
+func (f *FileWatcher) Start(config ConfigMap) error {
+	const errMessage = "failed to start watcher: %w"
+
+	var err error
+
+	if f.Loader == nil {
+		return fmt.Errorf(errMessage, ErrLoaderNotSet)
+	}
```

Comment:
die validation könnte man jetzt auch in die Init functions packen schätze ich, genau so könnte man auch ne validation auf den path machen damit der nicht leer ist, evtl sogar auch die optional funcs mit nem error versehen dann damit man auf der sicheren seite ist? 🤔